lib/mayaUsd/ufe/UsdTransform3dCommonAPI.cpp

```diff
@@ -24,167 +25,160 @@ namespace ufe {
+    extractXform(item, time, t, r, s, pvt);
+
+    return t;
+}
+
+GfVec3d extractRotation(const UsdSceneItem::Ptr& item, const UsdTimeCode& time)
+{
+    GfVec3d t;
+    GfVec3f r, s, pvt;
+    extractXform(item, time, t, r, s, pvt);
+
+    return r;
+}
+
+GfVec3d extractScale(const UsdSceneItem::Ptr& item, const UsdTimeCode& time)
+{
+    GfVec3d t;
+    GfVec3f r, s, pvt;
+    extractXform(item, time, t, r, s, pvt);
+
+    return s;
+}
+
+GfVec3d extractPivot(const UsdSceneItem::Ptr& item, const UsdTimeCode& time)
+{
+    GfVec3d t;
+    GfVec3f r, s, pvt;
+    extractXform(item, time, t, r, s, pvt);
+
+    return pvt;
+}
+
+// Common base class for the translation, rotation, scale and pivot commands.
+//
+// Only the handleSet() function, which sets the correct part of the transform
+// is done in the sub-classes.
+class UsdTRSUndoableCmdBase : public UsdUndoableCmdBase<Ufe::SetVector3dUndoableCommand>
 {
 public:
-    UsdTranslateUndoableCmd(const UsdSceneItem::Ptr& item, const UsdTimeCode& time)
-        : Ufe::TranslateUndoableCommand(item->path())
-        , _time(time)
+    UsdTRSUndoableCmdBase(
+        const GfVec3d&           value,
+        const UsdSceneItem::Ptr& item,
+        const UsdTimeCode&       time)
+        : UsdUndoableCmdBase<Ufe::SetVector3dUndoableCommand>(VtValue(value), item->path(), time)
         , _commonAPI(item->prim())
     {
-        GfVec3d                              t;
-        GfVec3f                              r, s, pvt;
-        UsdGeomXformCommonAPI::RotationOrder rotOrder;
-
-        if (!_commonAPI.GetXformVectorsByAccumulation(&t, &r, &s, &pvt, &rotOrder, time)) {
-            TF_FATAL_ERROR(
-                "Cannot read common API transform values for prim %s",
-                item->prim().GetPath().GetText());
-        }
-
-        _prevT = t;
-        _t = t;
     }
 
-    void undo() override { _commonAPI.SetTranslate(_prevT, _time); }
-    void redo() override { _commonAPI.SetTranslate(_t, _time); }
+    UsdTRSUndoableCmdBase(
+        const GfVec3f&           value,
+        const UsdSceneItem::Ptr& item,
+        const UsdTimeCode&       time)
+        : UsdUndoableCmdBase<Ufe::SetVector3dUndoableCommand>(VtValue(value), item->path(), time)
+        , _commonAPI(item->prim())
+    {
+    }
 
-    // Executes the command by setting the translation onto the transform op.
+    // Executes the command by setting the translation/rotation/scale/pivot into the value
+    // that will be used when we execute the transform op.
     bool set(double x, double y, double z) override
     {
-        _t = GfVec3d(x, y, z);
-        redo();
+        setNewValue(VtValue(GfVec3d(x, y, z)));
+        execute();
         return true;
     }
 
-private:
-    const UsdTimeCode     _time;
+protected:
     UsdGeomXformCommonAPI _commonAPI;
-    GfVec3d               _prevT, _t;
 };
 
-class UsdRotateUndoableCmd : public Ufe::RotateUndoableCommand
+class UsdTranslateUndoableCmd : public UsdTRSUndoableCmdBase
```

Comment:
All other changes are basically the same refactor of replacing the old base class and code logic to use the new base classes that handle most of the work. All that is kept is implementing handleSet() and sometimes set().